readme.txt
Setup:

In a command prompt, run "python setup.py install"

Arguments:
    -d: Display on default screen (not OLED screen)
    -k: Accept input using up, down, and q keys (only compatible when used with -d command)
    -c: Accept input from the command line

Dependencies list:
	* math
	* random
	* sys
	* matplotlib
	* numpy
	* yaml
	* pylab
	* scipy
	
Dependecies for OLED display:
	* RPi.GPIO
	* OLED_Driver (included in this repo)
	* PIL (pillow)
	
Dependencies for regular display: 
	* pygame
	* os
	* operator


changelog.txt
Version 5.0:

* Command-Line arguments added
* Increase size of display for non-OLED display

-------------------------------------------------------

Version 4.01:

* Option colors changed
* Selected option is outlined
* Circle color white, background black
* Added crosshair in center of circle
* Rephrased test question

-------------------------------------------------------

Version 4.0:

* Support for buttons

-------------------------------------------------------

Version 3.0:

* Support for display and custom driver
* Removed circle number/sanity check indicator (not enough space)
* Title no longer fades in (is complicated in new driver - will look into more later)

-------------------------------------------------------

Version 2.2:

* Incorporated bugfixes made by team (v2.1, see 3/13/19 notes))
* Removed code that centered program on screen
* Added code to display on correct screen

-------------------------------------------------------

Version 2.0:

General GUI:
* Re-sized program to 128x128
* Reformatted button select
* Changed background to grey
* Title fade in
* Enlarge text

Testing:
* answer after picture shown for 0.5 seconds
* removed circle background
* better circles (thiccer and smoother)
* shows current circle out of ten
* after testing, display test complete (removed profile)

Data:
* data is constantly exported to file in data/data

3/3/19
*bugfix for "distortions" in display.circle—Matlab code had an off by one error.
*Changed the line width to 2
*changed the color from grey to black
*added "ax.set_rmax(6)" to display.circle, same function as rlim([0 6]) in Matlab code
*changed initial threshold_difference to 15, and moved "threshold_difference = abs..." to the end
of the while loop. Small bugfix, but means that the program terminates when it should instead of
one iteration later.




button_test.py
import RPi.GPIO as GPIO  # Import Raspberry Pi GPIO library

GPIO.setwarnings(False)  # Ignore warning for now
GPIO.setmode(GPIO.BOARD)  # Use physical pin numbering
GPIO.setup(10, GPIO.IN,
           pull_up_down=GPIO.PUD_DOWN)  # Set pin 10 to be an input pin and set initial value to be pulled low (off)

while True:  # Run forever
    if GPIO.input(10) == GPIO.HIGH: print("Button was pushed!")


OLED_Driver.py
# -*- coding:UTF-8 -*-

import time

import RPi.GPIO as GPIO
import spidev

# SSD1351
SSD1351_WIDTH = 128
SSD1351_HEIGHT = 128
SSD1351_CMD_SETCOLUMN = 0x15
SSD1351_CMD_SETROW = 0x75
SSD1351_CMD_WRITERAM = 0x5C
SSD1351_CMD_READRAM = 0x5D
SSD1351_CMD_SETREMAP = 0xA0
SSD1351_CMD_STARTLINE = 0xA1
SSD1351_CMD_DISPLAYOFFSET = 0xA2
SSD1351_CMD_DISPLAYALLOFF = 0xA4
SSD1351_CMD_DISPLAYALLON = 0xA5
SSD1351_CMD_NORMALDISPLAY = 0xA6
SSD1351_CMD_INVERTDISPLAY = 0xA7
SSD1351_CMD_FUNCTIONSELECT = 0xAB
SSD1351_CMD_DISPLAYOFF = 0xAE
SSD1351_CMD_DISPLAYON = 0xAF
SSD1351_CMD_PRECHARGE = 0xB1
SSD1351_CMD_DISPLAYENHANCE = 0xB2
SSD1351_CMD_CLOCKDIV = 0xB3
SSD1351_CMD_SETVSL = 0xB4
SSD1351_CMD_SETGPIO = 0xB5
SSD1351_CMD_PRECHARGE2 = 0xB6
SSD1351_CMD_SETGRAY = 0xB8
SSD1351_CMD_USELUT = 0xB9
SSD1351_CMD_PRECHARGELEVEL = 0xBB
SSD1351_CMD_VCOMH = 0xBE
SSD1351_CMD_CONTRASTABC = 0xC1
SSD1351_CMD_CONTRASTMASTER = 0xC7
SSD1351_CMD_MUXRATIO = 0xCA
SSD1351_CMD_COMMANDLOCK = 0xFD
SSD1351_CMD_HORIZSCROLL = 0x96
SSD1351_CMD_STOPSCROLL = 0x9E
SSD1351_CMD_STARTSCROLL = 0x9F

# color
BLACK = 0x0000
BLUE = 0x001F
RED = 0xF800
GREEN = 0x07E0
CYAN = 0x07FF
MAGENTA = 0xF81F
YELLOW = 0xFFE0
WHITE = 0xFFFF
# buffer
color_byte = [0x00, 0x00]
color_fill_byte = [0x00, 0x00] * (SSD1351_WIDTH)

# GPIO Set
GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)
OLED_RST_PIN = 25
OLED_DC_PIN = 24
OLED_CS_PIN = 8
GPIO.setup(OLED_RST_PIN, GPIO.OUT)
GPIO.setup(OLED_DC_PIN, GPIO.OUT)
GPIO.setup(OLED_CS_PIN, GPIO.OUT)
# GPIO init
GPIO.setwarnings(False)
GPIO.setup(OLED_RST_PIN, GPIO.OUT)
GPIO.setup(OLED_DC_PIN, GPIO.OUT)
GPIO.setup(OLED_CS_PIN, GPIO.OUT)
# SPI init
SPI = spidev.SpiDev(0, 0)
SPI.max_speed_hz = 9000000
SPI.mode = 0b00


def Set_Color(color):
    color_byte[0] = (color >> 8) & 0xff
    color_byte[1] = color & 0xff


def OLED_RST(x):
    if x == 1:
        GPIO.output(OLED_RST_PIN, GPIO.HIGH)
    elif x == 0:
        GPIO.output(OLED_RST_PIN, GPIO.LOW)


def OLED_DC(x):
    if x == 1:
        GPIO.output(OLED_DC_PIN, GPIO.HIGH)
    elif x == 0:
        GPIO.output(OLED_DC_PIN, GPIO.LOW)


def OLED_CS(x):
    if x == 1:
        GPIO.output(OLED_CS_PIN, GPIO.HIGH)
    elif x == 0:
        GPIO.output(OLED_CS_PIN, GPIO.LOW)


def SPI_WriteByte(byte):
    SPI.writebytes(byte)


def Write_Command(cmd):
    OLED_CS(0)
    OLED_DC(0)
    SPI_WriteByte([cmd])
    OLED_CS(1)


def Write_Data(dat):
    OLED_CS(0)
    OLED_DC(1)
    SPI_WriteByte([dat])
    OLED_CS(1)


def Write_Datas(data):
    OLED_CS(0)
    OLED_DC(1)
    SPI_WriteByte(data)
    OLED_CS(1)


def RAM_Address():
    Write_Command(0x15)
    Write_Data(0x00)
    Write_Data(0x7f)
    Write_Command(0x75)
    Write_Data(0x00)
    Write_Data(0x7f)


def Fill_Color(color):
    RAM_Address()
    Write_Command(0x5c)
    Set_Color(color)
    color_fill_byte = color_byte * SSD1351_WIDTH
    OLED_CS(0)
    OLED_DC(1)
    for i in range(0, SSD1351_HEIGHT):
        SPI_WriteByte(color_fill_byte)
    OLED_CS(1)


def Clear_Screen():
    RAM_Address()
    Write_Command(0x5c)
    color_fill_byte = [0x00, 0x00] * SSD1351_WIDTH
    OLED_CS(0)
    OLED_DC(1)
    for i in range(0, SSD1351_HEIGHT):
        SPI_WriteByte(color_fill_byte)
    OLED_CS(1)


def Draw_Pixel(x, y):
    # Bounds check.
    if ((x >= SSD1351_WIDTH) or (y >= SSD1351_HEIGHT)):
        return
    if ((x < 0) or (y < 0)):
        return
    Set_Address(x, y)
    # transfer data
    Write_Datas(color_byte)


def Set_Coordinate(x, y):
    if ((x >= SSD1351_WIDTH) or (y >= SSD1351_HEIGHT)):
        return
    # Set x and y coordinate
    Write_Command(SSD1351_CMD_SETCOLUMN)
    Write_Data(x)
    Write_Data(SSD1351_WIDTH - 1)
    Write_Command(SSD1351_CMD_SETROW)
    Write_Data(y)
    Write_Data(SSD1351_HEIGHT - 1)
    Write_Command(SSD1351_CMD_WRITERAM)


def Set_Address(column, row):
    Write_Command(SSD1351_CMD_SETCOLUMN)
    Write_Data(column)  # X start
    Write_Data(column)  # X end
    Write_Command(SSD1351_CMD_SETROW)
    Write_Data(row)  # Y start
    Write_Data(row + 7)  # Y end
    Write_Command(SSD1351_CMD_WRITERAM)


def Write_text(dat):
    for i in range(0, 8):
        if (dat & 0x01):
            Write_Datas(color_byte)
        else:
            Write_Datas([0x00, 0x00])
        dat = dat >> 1


def Invert(v):
    if (v):
        Write_Command(SSD1351_CMD_INVERTDISPLAY)
    else:
        Write_Command(SSD1351_CMD_NORMALDISPLAY)


def Draw_Pixel(x, y):
    # Bounds check.
    if ((x >= SSD1351_WIDTH) or (y >= SSD1351_HEIGHT)):
        return
    if ((x < 0) or (y < 0)):
        return
    Set_Address(x, y)
    # transfer data
    Write_Datas(color_byte)


def Delay(x):
    time.sleep(x / 1000.0)


def Device_Init():
    OLED_CS(0)
    OLED_RST(0)
    Delay(500)
    OLED_RST(1)
    Delay(500)

    Write_Command(0xfd)  # command lock
    Write_Data(0x12)
    Write_Command(0xfd)  # command lock
    Write_Data(0xB1)

    Write_Command(0xae)  # display off
    Write_Command(0xa4)  # Normal Display mode

    Write_Command(0x15)  # set column address
    Write_Data(0x00)  # column address start 00
    Write_Data(0x7f)  # column address end 95
    Write_Command(0x75)  # set row address
    Write_Data(0x00)  # row address start 00
    Write_Data(0x7f)  # row address end 63

    Write_Command(0xB3)
    Write_Data(0xF1)

    Write_Command(0xCA)
    Write_Data(0x7F)

    Write_Command(0xa0)  # set re-map & data format
    Write_Data(0x74)  # Horizontal address increment

    Write_Command(0xa1)  # set display start line
    Write_Data(0x00)  # start 00 line

    Write_Command(0xa2)  # set display offset
    Write_Data(0x00)

    Write_Command(0xAB)
    Write_Command(0x01)

    Write_Command(0xB4)
    Write_Data(0xA0)
    Write_Data(0xB5)
    Write_Data(0x55)

    Write_Command(0xC1)
    Write_Data(0xC8)
    Write_Data(0x80)
    Write_Data(0xC0)

    Write_Command(0xC7)
    Write_Data(0x0F)

    Write_Command(0xB1)
    Write_Data(0x32)

    Write_Command(0xB2)
    Write_Data(0xA4)
    Write_Data(0x00)
    Write_Data(0x00)

    Write_Command(0xBB)
    Write_Data(0x17)

    Write_Command(0xB6)
    Write_Data(0x01)

    Write_Command(0xBE)
    Write_Data(0x05)

    Write_Command(0xA6)

    Clear_Screen()
    Write_Command(0xaf)


# Draw a horizontal line ignoring any screen rotation.
def Draw_FastHLine(x, y, length):
    # Bounds check
    if ((x >= SSD1351_WIDTH) or (y >= SSD1351_HEIGHT)):
        return
    # X bounds check
    if ((x + length) > SSD1351_WIDTH):
        length = SSD1351_WIDTH - x - 1
    if (length < 0):
        return
    # set location
    Write_Command(SSD1351_CMD_SETCOLUMN)
    Write_Data(x)
    Write_Data(x + length - 1)
    Write_Command(SSD1351_CMD_SETROW)
    Write_Data(y)
    Write_Data(y)
    # fill!
    Write_Command(SSD1351_CMD_WRITERAM)

    for i in range(0, length):
        Write_Datas(color_byte)


def Draw_FastVLine(x, y, length):
    # Bounds check
    if ((x >= SSD1351_WIDTH) or (y >= SSD1351_HEIGHT)):
        return
    # X bounds check
    if (y + length > SSD1351_HEIGHT):
        length = SSD1351_HEIGHT - y - 1
    if (length < 0):
        return
    # set location
    Write_Command(SSD1351_CMD_SETCOLUMN)
    Write_Data(x)
    Write_Data(x)
    Write_Command(SSD1351_CMD_SETROW)
    Write_Data(y)
    Write_Data(y + length - 1)
    # fill!
    Write_Command(SSD1351_CMD_WRITERAM)

    for i in range(0, length):
        Write_Datas(color_byte)


def Display_Image(Image):
    if (Image == None):
        return

    Set_Coordinate(0, 0)
    buffer1 = Image.load()
    for j in range(0, SSD1351_WIDTH):
        for i in range(0, SSD1351_HEIGHT):
            color_fill_byte[i * 2] = ((buffer1[i, j][0] & 0xF8) | (buffer1[i, j][1] >> 5))
            color_fill_byte[i * 2 + 1] = (((buffer1[i, j][1] << 3) & 0xE0) | (buffer1[i, j][2] >> 3))
        Write_Datas(color_fill_byte)


kalEYEdoscope.py
import math
import random
import sys

import matplotlib

matplotlib.use('pdf')
import matplotlib.pyplot as plt
import numpy as np
import yaml
from pylab import sin

# DEFINE CONSTANT VARIABLES

NUMCIRCLES = 10
WIDTH = 128
HEIGHT = 128
BUTTON1 = 14
BUTTON2 = 15
INTYPE = "BUTTON"
DISPLAYTYPE = "OLED"

# Interpret command-line arguments
if len(sys.argv) > 1:
    if "-d" in sys.argv:
        DISPLAYTYPE = "HDMI"
        WIDTH = 384
        HEIGHT = 384
        import pygame
        import os
        import operator
    else:
        import RPi.GPIO as GPIO
        import OLED_Driver as OLED
        import Image
        import ImageDraw
        import ImageFont
    if "-c" in sys.argv:
        INTYPE = "SHELL"
    if "-k" in sys.argv:
        if DISPLAYTYPE == "OLED":
            print("Error, keyboard must be used with default computer display")
            exit(1)
        INTYPE = "KEYBOARD"


# DEFINE CONSTANT FUNCTIONS

def get_input():
    quit_count = 0
    if INTYPE == "BUTTON":
        while True:
            if GPIO.input(BUTTON1) == GPIO.HIGH and GPIO.input(BUTTON2) == GPIO.HIGH:
                if quit_count == 2:
                    program_quit()
                OLED.Delay(1000)
                quit_count += 1
            elif GPIO.input(BUTTON1) == GPIO.HIGH:
                return 1
            elif GPIO.input(BUTTON2) == GPIO.HIGH:
                return 2
    elif INTYPE == "SHELL":
        while True:
            if sys.version_info[0] < 3:
                i = raw_input("input: ")
            else:
                i = input("input: ")
            if i == "12": program_quit()
            if i == "1": return 1
            if i == "2": return 2
    elif INTYPE == "KEYBOARD":
        while True:
            for e in pygame.event.get():
                if e.type == pygame.KEYDOWN:
                    if e.key == pygame.K_UP: return 1
                    if e.key == pygame.K_DOWN: return 2
                    if e.key == pygame.K_q: program_quit()


def clear_screen():
    """
    Clears the screen by filling with background
    """
    if DISPLAYTYPE == "OLED":
        OLED.Clear_Screen()
        clear_image = Image.new("RGB", (OLED.SSD1351_WIDTH, OLED.SSD1351_HEIGHT), "DIMGREY")
        clear_draw = ImageDraw.Draw(clear_image)
        return clear_draw, clear_image
    elif DISPLAYTYPE == "HDMI":
        screen.fill(color_background)
        return None, None


def update(update_image=None):
    if DISPLAYTYPE == "OLED":
        OLED.Display_Image(update_image)
    elif DISPLAYTYPE == "HDMI":
        pygame.display.flip()


def update_data():
    """
    Saves current data to file
    """
    with open('data/data', "w") as save_data:
        yaml.dump(data, save_data)


def update_buttons(left, right, title=None, subtitle=None, update_draw=None, update_image=None):
    """
    Updates the text displayed for each button instruction
    :param update_image:
    :param update_draw:
    :param subtitle: text to be displayed just below title
    :param title: text to be displayed above option buttons
    :param left: text to be displayed for the left button
    :param right: text to be displayed for the right button
    """
    if DISPLAYTYPE == "OLED":
        right_text = update_draw.textsize(right, font=font_normal)
        left_text = update_draw.textsize(left, font=font_normal)

        update_draw.rectangle([(WIDTH // 2 - right_text[0] // 2 - padding, HEIGHT // 2),
                               (WIDTH // 2 + right_text[0] // 2 + padding, HEIGHT // 2 + right_text[1] + padding)],
                              fill="BLUE")

        update_draw.rectangle([(WIDTH // 2 - left_text[0] // 2 - padding, HEIGHT * 3 // 4),
                               (WIDTH // 2 + left_text[0] // 2 + padding, HEIGHT * 3 // 4 + left_text[1] + padding)],
                              fill="DARKORANGE")

        update_draw.text((WIDTH // 2 - right_text[0] // 2, HEIGHT // 2), right, font=font_normal)
        update_draw.text((WIDTH // 2 - left_text[0] // 2, HEIGHT * 3 // 4), left, font=font_normal)

        if title:
            title_text = update_draw.textsize(title, font_subtitle)
            update_draw.text((WIDTH // 2 - title_text[0] // 2, padding), title, font=font_subtitle)

        if subtitle:
            subtitle_text = update_draw.textsize(subtitle, font_subtitle)
            update_draw.text((WIDTH // 2 - subtitle_text[0] // 2, subtitle_text[1] + padding * 2), subtitle,
                             font=font_subtitle)

        update(update_image)
        button = get_input()
        if button == 1:
            update_draw.rectangle([(WIDTH // 2 - right_text[0] // 2 - padding, HEIGHT // 2),
                                   (WIDTH // 2 + right_text[0] // 2 + padding, HEIGHT // 2 + right_text[1] + padding)],
                                  outline="BLACK")
            update(update_image)
            return 1
        if button == 2:
            update_draw.rectangle([(WIDTH // 2 - left_text[0] // 2 - padding, HEIGHT * 3 // 4),
                                   (
                                       WIDTH // 2 + left_text[0] // 2 + padding,
                                       HEIGHT * 3 // 4 + left_text[1] + padding)],
                                  outline="BLACK")
            update(update_image)
            return 2

    elif DISPLAYTYPE == "HDMI":
        right_text = font_normal.render(right, True, color_white)
        left_text = font_normal.render(left, True, color_white)
        right_surface = pygame.Surface(tuple(map(operator.add, font_normal.size(right), (5, 5))))
        left_surface = pygame.Surface(tuple(map(operator.add, font_normal.size(left), (7, 7))))
        right_surface.fill(color_background)
        left_surface.fill(color_background)

        pygame.draw.rect(right_surface, color_right, right_surface.get_rect())
        pygame.draw.rect(left_surface, color_left, left_surface.get_rect())

        right_surface.blit(right_text, (right_surface.get_width() // 2 - right_text.get_width() // 2,
                                        right_surface.get_height() // 2 - right_text.get_height() // 2))
        left_surface.blit(left_text, (left_surface.get_width() // 2 - left_text.get_width() // 2,
                                      left_surface.get_height() // 2 - left_text.get_height() // 2))

        screen.blit(right_surface, (WIDTH // 2 - right_surface.get_width() // 2, HEIGHT // 2))
        screen.blit(left_surface, (WIDTH // 2 - left_surface.get_width() // 2, HEIGHT * 3 // 4))

        if title:
            title_text = font_title.render(title, True, color_white)
            screen.blit(title_text, (WIDTH // 2 - title_text.get_width() // 2, title_text.get_height()))

        if subtitle:
            subtitle_text = font_title.render(subtitle, True, color_white)
            screen.blit(subtitle_text, (WIDTH // 2 - subtitle_text.get_width() // 2,
                                        subtitle_text.get_height() * 2 + padding))

        update()
        button = get_input()
        if button == 1:
            pygame.draw.rect(right_surface, color_black, right_surface.get_rect(), 2)
            right_surface.blit(right_text, (right_surface.get_width() // 2 - right_text.get_width() // 2,
                                            right_surface.get_height() // 2 - right_text.get_height() // 2))
            screen.blit(right_surface, (WIDTH // 2 - right_surface.get_width() // 2, HEIGHT // 2))

            update()
            pygame.time.delay(100)
            return 1
        if button == 2:
            pygame.draw.rect(left_surface, color_black, left_surface.get_rect(), 2)
            left_surface.blit(left_text, (left_surface.get_width() // 2 - left_text.get_width() // 2,
                                          left_surface.get_height() // 2 - left_text.get_height() // 2))
            screen.blit(left_surface, (WIDTH // 2 - left_surface.get_width() // 2, HEIGHT * 3 // 4))

            update()
            pygame.time.delay(100)
            return 2


# INITIALIZATION

# Start display

if DISPLAYTYPE == "OLED":
    OLED.Device_Init()
elif DISPLAYTYPE == "HDMI":
    os.environ['SDL_VIDEO_CENTERED'] = '1'
    pygame.init()
    size = WIDTH, HEIGHT
    pygame.display.set_caption("KalEYEdoscope")
    screen = pygame.display.set_mode(size, pygame.NOFRAME)
    clock = pygame.time.Clock()

# Initialize buttons

if INTYPE == "BUTTON":
    GPIO.setup(BUTTON1, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
    GPIO.setup(BUTTON2, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)

# Load frequently used variables

data = yaml.safe_load(open("data/data"))
state = "Start"

# Fonts
if DISPLAYTYPE == "OLED":
    font_title = ImageFont.truetype('cambriab.ttf', 24)
    font_subtitle = ImageFont.truetype('cambriab.ttf', 16)
    font_normal = ImageFont.truetype('cambriab.ttf', 18)
else:
    font_title = pygame.font.Font(None, 66)
    font_subtitle = pygame.font.Font(None, 48)
    font_normal = pygame.font.Font(None, 66)

# Spacings
padding = 2

# Colors
color_black = (0, 0, 0)
color_gray = (100, 100, 100)
color_white = (255, 255, 255)
color_background = (105, 105, 105)
color_right = (0, 0, 255)
color_left = (255, 140, 0)


# ACTIVITY FUNCTIONS

def start():
    """"
    Main screen event.
    """
    global state
    if DISPLAYTYPE == "OLED":
        draw, image = clear_screen()
        welcome = "Welcome to your"
        title = "KalEYEdoscope"
        welcome_text = draw.textsize(welcome, font_subtitle)
        title_text = draw.textsize(title, font_subtitle)
        draw.text((WIDTH // 2 - welcome_text[0] // 2, welcome_text[1]), welcome, font=font_subtitle, fill="BLACK")
        draw.text((WIDTH // 2 - title_text[0] // 2, welcome_text[1] * 2 + padding), title, font=font_subtitle,
                  fill="BLACK")

        button = update_buttons("Exit", "Start", update_draw=draw, update_image=image)

    else:
        clear_screen()
        text = font_title.render("Welcome to your", True, color_white)
        surface = pygame.Surface(font_title.size("Welcome to your"))
        surface.fill(color_background)
        surface.blit(text, (0, 0))
        text2 = font_title.render("KalEYEdoscope", True, color_white)
        surface2 = pygame.Surface(font_title.size("KalEYEdoscope"))
        surface2.fill(color_background)
        surface2.blit(text2, (0, 0))
        screen.blit(surface, (WIDTH // 2 - text.get_width() // 2, text.get_height()))
        screen.blit(surface2, (WIDTH // 2 - text2.get_width() // 2, text.get_height() * 2 + padding))
        update()
        button = update_buttons("Exit", "Start")
        update()

    if button == 1:
        if data["new_user"]:
            state = "Baseline"
        else:
            state = "Test"

    else:
        program_quit()


def baseline():
    """
    Record a baseline test event.
    """
    global state
    draw, image = clear_screen()

    data["new_user"] = False
    data["tests"] = {"0": None}
    data["number_of_tests"] = -1
    update_data()

    update_buttons("Start", "Start", "Record a ", "baseline.", draw, image)
    state = "Test"


def display_circle(angle, theta, r):
    """
    Displays a circle for half a second, then prompts the user to answer whether the circle is distorted.
    :param angle: angle of the circle
    :param theta: theta for the circle
    :param r: circle radius
    :return: user's selection
    """
    tan = math.tan(math.radians(angle))
    if tan != 0:
        slope = 1 / tan
    else:
        slope = 99999
    b = random.randint(5, 8)
    phi = random.random() * 2 * np.pi
    distortions = [r / (slope * 5), r / (slope * 6), r / (slope * 7), r / (slope * 8)]
    rho = r + distortions[b - 5] * sin(b * (theta + phi))
    plt.polar(theta, rho, linewidth=2, color='w')
    ax = plt.gca()
    ax.grid(False)
    ax.set_rmax(6)
    ax.set_yticklabels([])
    ax.set_xticklabels([])
    ax.spines['polar'].set_visible(False)
    plt.gcf().savefig('images/current_circle.png', bbox_inches='tight', dpi=(WIDTH // 4), transparent=True,
                      facecolor='k')
    plt.close(plt.gcf())

    if DISPLAYTYPE == "OLED":
        circle_image = Image.open("images/current_circle.png", mode='r')
        draw = ImageDraw.Draw(circle_image)
        # Center cross hair
        draw.line((WIDTH // 2, HEIGHT // 2 - 2, WIDTH // 2, HEIGHT // 2 + 2))
        draw.line((WIDTH // 2 - 2, HEIGHT // 2, WIDTH // 2 + 2, HEIGHT // 2))

        OLED.Clear_Screen()
        update(circle_image)
        OLED.Delay(500)
    elif DISPLAYTYPE == "HDMI":
        circle = pygame.image.load("images/current_circle.png")
        clear_screen()
        screen.blit(circle, (WIDTH // 2 - circle.get_width() // 2,
                             HEIGHT // 2 - circle.get_height() // 2))
        pygame.draw.line(screen, color_white, (WIDTH // 2, HEIGHT // 2 - 5), (WIDTH // 2, HEIGHT // 2 + 5))
        pygame.draw.line(screen, color_white, (WIDTH // 2 - 5, HEIGHT // 2), (WIDTH // 2 + 5, HEIGHT // 2))
        update()
        pygame.time.delay(500)
        clear_screen()

    draw, image = clear_screen()
    button = update_buttons("No", "Yes", "Was that a", "perfect circle?", draw, image)

    # Note that choice returns "y" for a distorted circle and "n" for a normal circle
    if button == 1:
        choice = "n"
    elif button == 2:
        choice = "y"
    else:
        choice = "error"

    return choice


def test():
    """
    Test event
    """
    global state

    # Pick eye
    draw, image = clear_screen()
    button = update_buttons("Left", "Right", "Select an eye", "to test.", draw, image)
    eye = "Error"

    if button == 1:
        eye = "L"
    elif button == 2:
        eye = "R"

    # Read user data
    testnum = str(data["number_of_tests"] + 1)
    data["tests"][testnum] = {}
    data["number_of_tests"] = data["number_of_tests"] + 1
    data["tests"][testnum]["eye"] = eye

    # Calculate Circle Parameters
    upper_threshold = [15]
    lower_threshold = [0]
    current_test = 0
    threshold_difference = 15
    convergence = 0.372
    theta = np.arange(0, 4 * np.pi, 0.01)[1:]
    r = random.randint(3, 5)
    incorrect_sanity_checks = 0
    num_sanity_checks = 0
    while threshold_difference >= 0.75:
        current_big_t = upper_threshold[current_test]
        current_small_t = lower_threshold[current_test]
        # threshold_difference = abs(current_big_t - current_small_t)
        if random.randint(0, 100) <= 85 or current_test == 0:
            # Normal Test
            angle = current_big_t - convergence * threshold_difference
            choice = display_circle(angle, theta, r)
            if choice == "y":
                upper_threshold.append(angle)
            else:
                upper_threshold.append(
                    (current_big_t + min(np.median(upper_threshold), np.mean(upper_threshold))) / 2
                )
            angle = current_small_t + convergence * abs(upper_threshold[current_test + 1] - current_small_t)
            choice = display_circle(angle, theta, r)
            if choice == "y":
                lower_threshold.append(
                    (current_small_t + min(np.median(lower_threshold), np.mean(lower_threshold))) / 2
                )
            else:
                lower_threshold.append(angle)
        else:
            # Sanity Check Test
            num_sanity_checks += 1
            angle = random.choice([current_big_t, current_small_t])
            choice = display_circle(angle, theta, r)
            if choice == "y":
                upper_threshold.append(current_big_t)
                if angle == current_small_t:
                    lower_threshold.append(lower_threshold[current_test - 1])
                    incorrect_sanity_checks += 1
                else:
                    lower_threshold.append(current_small_t)
            elif choice == "n":
                lower_threshold.append(current_small_t)
                if angle == current_small_t:
                    upper_threshold.append(upper_threshold[current_test - 1])
                    incorrect_sanity_checks += 1
                else:
                    upper_threshold.append(current_big_t)

        current_test += 1
        current_big_t = upper_threshold[current_test]
        current_small_t = lower_threshold[current_test]
        threshold_difference = abs(current_big_t - current_small_t)

    data["tests"][testnum]["lower_thresholds"] = [float(x) for x in lower_threshold]
    data["tests"][testnum]["upper_thresholds"] = [float(x) for x in upper_threshold]
    data["tests"][testnum]["check_sanity_num"] = num_sanity_checks
    data["tests"][testnum]["incorrect_sanity_checks"] = incorrect_sanity_checks
    update_data()

    draw, image = clear_screen()
    update_buttons("Start Menu", "Start Menu", "Test Completed", draw, image)
    state = "Start"


def program_quit():
    if DISPLAYTYPE == "OLED":
        OLED.Clear_Screen()
        GPIO.cleanup()
        exit()
    else:
        sys.exit()


# MAIN RUNNING LOOP
while True:
    if state == "Start":
        start()

    elif state == "Baseline":
        baseline()

    elif state == "Test":
        test()

    elif state == "Quit":
        program_quit()

    else:
        print(state)
        print(" not implemented yet\n")
        state = "Quit"
